src/providers/azure.rs

```diff
@@ -0,0 +1,519 @@
+
+        azcmd::create_resource_group(region, &rg_name)?;
+
+        Ok(Self {
+            log: Some(log),
+            region,
+            resource_group_name: rg_name,
+            machines: vec![],
+        })
+    }
+}
+
+impl super::Launcher for RegionLauncher {
+    type MachineDescriptor = Setup;
+
+    fn launch(&mut self, l: super::LaunchDescriptor<Self::MachineDescriptor>) -> Result<(), Error> {
+        self.log = Some(l.log);
+        let log = self.log.as_ref().unwrap();
+        let max_wait = l.max_wait;
+        let vms: Result<Vec<(String, String, _)>, Error> = l.machines
+            .into_iter()
+            .map(|(nickname, desc)| {
+                let vm_name = super::rand_name_sep("vm", "-");
+                debug!(log, "setting up azure instance"; "nickname" => &nickname, "vm_name" => &vm_name);
+
+                let pub_ip = azcmd::create_vm(
+                    &self.resource_group_name,
+                    &vm_name,
+                    &desc.instance_type,
+                    &desc.image,
+                    &desc.username,
+                )?;
+
+                azcmd::open_ports(&self.resource_group_name, &vm_name)?;
+
+                Ok((nickname, pub_ip, desc))
+            }).collect();
+
+        use rayon::prelude::*;
+        self.machines = vms?
+            .into_par_iter()
+            .map(|(nickname, pub_ip, desc)| {
+                if let Setup {
+                    ref username,
+                    setup_fn: Some(ref f),
+                    ..
+                } = desc
+                {
+                    super::setup_machine(
+                        log,
+                        &nickname,
+                        &pub_ip,
+                        &username,
+                        max_wait,
+                        None,
+                        f.as_ref(),
+                    )?;
+                }
+
+                Ok(Descriptor {
+                    name: nickname,
+                    username: desc.username,
+                    ip: pub_ip,
+                })
+            })
+            .collect::<Result<Vec<Descriptor>, Error>>()?;
+
+        Ok(())
+    }
+
+    fn connect_all<'l>(&'l self) -> Result<HashMap<String, crate::Machine<'l>>, Error> {
+        let log = self.log.as_ref().expect("RegionLauncher uninitialized");
+        self.machines
+            .iter()
+            .map(|desc| {
+                let Descriptor { name, username, ip } = desc;
+                let mut m = crate::Machine {
+                    nickname: name.clone(),
+                    public_dns: ip.clone(),
+                    public_ip: ip.clone(),
+                    ssh: None,
+                    _tsunami: Default::default(),
+                };
+
+                m.connect_ssh(log, username, None)?;
+                Ok((name.clone(), m))
+            })
+            .collect()
+    }
+}
+
+impl Drop for RegionLauncher {
+    fn drop(&mut self) {
```

Comment:
Same as for AWS, we should probably document this behavior somewhere.

Reply:
It is: https://github.com/jonhoo/tsunami/pull/22/files/6ca63d29e5c6243882bca7d3bb5fca31a53c042b#diff-4fc8e23f1c61397b75b19c2108b2c510R165

I will add for AWS though.